ui/components-react/src/components-react/propertygrid/MergingPropertyDataProvider.ts

```diff
@@ -93,12 +110,71 @@ export class MergingPropertyDataProvider implements IPropertyDataProvider {
   }
 
   private mergeCategories(dataArray: PropertyData[]): PropertyCategory[] {
-    const mergedCategories: PropertyCategory[] = [];
+    const mergedCategories = new Array<PropertyCategory>();
     dataArray.forEach((data) => {
-      mergedCategories.push(...data.categories);
+      for (const category of data.categories) {
+        const mergedCategoryIndex = mergedCategories.findIndex(
+          (mergedCat) => mergedCat.name === category.name
+        );
+        if (mergedCategoryIndex === -1) {
+          mergedCategories.push(category);
+          continue;
+        }
+
+        const newChildCategories = this.mergeChildCategories(
+          mergedCategories[mergedCategoryIndex].childCategories,
+          category.childCategories
+        );
+        mergedCategories[mergedCategoryIndex] = {
+          ...mergedCategories[mergedCategoryIndex],
+          ...(newChildCategories
+            ? { childCategories: [...newChildCategories] }
+            : {}),
+        };
```

Comment:
This almost duplicates `mergeChildCategories`. I think we could just do this:
- in the first iteration, simply copy source categories to merged categories list,
- in all others - use `mergeChildCategories` to create a new merged list.

Reply:
changed `mergeCategories` to this:
```
let mergedCategories: PropertyCategory[] | undefined;
if (dataArray.length > 0) {
  mergedCategories = [...dataArray[0].categories];
}

for (let i = 1; i < dataArray.length; ++i) {
  mergedCategories = this.mergeChildCategories(
    mergedCategories,
    dataArray[i].categories
  );
}

return mergedCategories ?? [];
```